File: cli/tbast.c
/****************************************************************************
** (c) Copyright IBM Corp. 2007 All rights reserved.
** 
** The following sample of source code ("Sample") is owned by International 
** Business Machines Corporation or one of its subsidiaries ("IBM") and is 
** copyrighted and licensed, not sold. You may use, copy, modify, and 
** distribute the Sample in any form without payment to IBM, for the purpose of 
** assisting you in the development of your applications.
** 
** The Sample code is provided to you on an "AS IS" basis, without warranty of 
** any kind. IBM HEREBY EXPRESSLY DISCLAIMS ALL WARRANTIES, EITHER EXPRESS OR 
** IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
** MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. Some jurisdictions do 
** not allow for the exclusion or limitation of implied warranties, so the above 
** limitations or exclusions may not apply to you. IBM shall not be liable for 
** any damages you suffer as a result of using, copying, modifying or 
** distributing the Sample, even if IBM has been advised of the possibility of 
** such damages.
*****************************************************************************
**
** SOURCE FILE NAME: tbast.c                                      
**                                                                        
** SAMPLE:  How to use staging table for updating deferred AST 
**          
**         This sample:
**         1. Creates a refresh-deferred summary table 
**         2. Creates a staging table for this summary table 
**         3. Applies contents of staging table to AST
**         4. Restores the data in a summary table  
**
** CLI FUNCTIONS USED:
**         SQLAllocHandle -- Allocate Handle
**         SQLExecDirect -- Execute a Statement Directly
**         SQLFetch -- Fetch Next Row
**         SQLFreeHandle -- Free Handle Resources
**         SQLSetConnectAttr -- Set Connection Attributes
**         SQLGetData -- Get Data From a Column                
**
** OUTPUT FILE: tbast.out (available in the online documentation)
*****************************************************************************
**
** For more information on the sample programs, see the README file.
**
** For information on developing CLI applications, see the CLI Guide
** and Reference.
**
** For information on using SQL statements, see the SQL Reference.
**
** For the latest information on programming, building, and running DB2 
** applications, visit the DB2 application development website: 
**     http://www.software.ibm.com/data/db2/udb/ad
****************************************************************************/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sqlcli1.h>
#include "utilcli.h" /* header file for CLI sample code */

int CreateStagingTable(SQLHANDLE);
int PropagateStagingToAst(SQLHANDLE);
int RestoreSummaryTable(SQLHANDLE);
int DisplayTable(SQLHANDLE,char *);
int DropTables(SQLHANDLE);

int main(int argc, char *argv[])
{
  SQLRETURN cliRC = SQL_SUCCESS;
  int rc = 0;
  SQLHANDLE henv; /* environment handle */
  SQLHANDLE hdbc; /* connection handles */

  char dbAlias[SQL_MAX_DSN_LENGTH + 1];
  char user[MAX_UID_LENGTH + 1];
  char pswd[MAX_PWD_LENGTH + 1];

  /* checks the command line arguments */
  rc = CmdLineArgsCheck1(argc, argv, dbAlias, user, pswd);
  if (rc != 0)
  {
    return rc;
  }

  printf("\nTHIS SAMPLE SHOWS THE USAGE OF STAGING TABLE TO UPDATE"); 
  printf("\nREFRESH DEFERRED AST AND RESTORE DATA IN A SUMMARY TABLE\n");
  printf("\n-----------------------------------------------------------\n");

  /* initialize the CLI application by calling a helper
     utility function defined in utilcli.c */
  rc = CLIAppInit(dbAlias,
                  user,
                  pswd,
                  &henv,
                  &hdbc,
                  (SQLPOINTER)SQL_AUTOCOMMIT_ON);
  if (rc != 0)
  {
    return rc;
  }
  
  /* create a base table, summary table, staging table */
  rc = CreateStagingTable(hdbc);
  if (rc != 0)
  {
    return rc;
  }
 
  printf("\n-----------------------------------------------------------\n");

  /* to show the propagation of changes of base table to
     summary tables through the staging table */
  printf("To show the propagation of changes from base table to\n");
  printf("summary tables through the staging table:\n");

  rc = PropagateStagingToAst(hdbc);
  if (rc != 0)
  {
    return rc;
  }

  printf("\n------------------------------------------------------------\n");
  
  /* to show restoring of data in a summary table */

  printf("\nTo show restoring of data in a summary table\n");
  rc = RestoreSummaryTable(hdbc);
  if (rc != 0)
  {
    return rc;
  }

  /* drop the created tables */
  printf("\nDrop the created tables\n");

  rc = DropTables(hdbc);
  if (rc != 0)
  {
    return rc;
  }

  /* terminate the CLI application by calling a helper
     utility function defined in utilcli.c */
  rc = CLIAppTerm(&henv, &hdbc, dbAlias);

  return rc;
} /* main */

/* creates base table, summary table and staging table */
int CreateStagingTable(SQLHANDLE hdbc)
{
  SQLRETURN cliRC = SQL_SUCCESS;
  int rc = 0;
  SQLHANDLE hstmt; /* statement handle */
  
  SQLCHAR *stmt1 = (SQLCHAR *)"CREATE TABLE T (C1 SMALLINT NOT NULL, "
                              "                C2 SMALLINT NOT NULL, "
                              "                C3 SMALLINT, C4 SMALLINT)";
  
  SQLCHAR *stmt2 = (SQLCHAR *)"CREATE SUMMARY TABLE D_AST AS "
                              "(SELECT C1, C2, COUNT(*) AS COUNT "
                              "FROM T GROUP BY C1, C2) DATA "
                              "INITIALLY DEFERRED REFRESH DEFERRED"; 

  SQLCHAR *stmt3 = (SQLCHAR *)"CREATE TABLE G FOR D_AST PROPAGATE "
                              "IMMEDIATE";

  printf("\nUSE THE CLI FUNCTIONS\n");
  printf("  SQLSetConnectAttr\n");
  printf("  SQLAllocHandle\n");
  printf("  SQLExecDirect\n");
  printf("  SQLFreeHandle\n");
  printf("TO CREATE BASE TABLE, SUMMARY TABLE, STAGING TABLE:\n");

  printf("\nCreating the base table T\n");
 
  /* set AUTOCOMMIT on */
  cliRC = SQLSetConnectAttr(hdbc,
                            SQL_ATTR_AUTOCOMMIT,
                            (SQLPOINTER)SQL_AUTOCOMMIT_ON,
                            SQL_NTS);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  printf("\n  Directly execute the statement\n");
  printf("    CREATE TABLE T\n");
  printf("      (C1 SMALLINT NOT NULL, C2 SMALLINT NOT NULL, \n");
  printf("       C3 SMALLINT, C4 SMALLINT)\n");
  
  /* create base table  */
  cliRC = SQLExecDirect(hstmt, stmt1, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* create summary table */
  printf("\nCreating summary table D_AST\n");
  printf("\n  Directly execute the statement\n");
  printf("    CREATE SUMMARY TABLE D_AST AS (SELECT C1, C2, COUNT(*)\n"); 
  printf("    AS COUNT FROM T GROUP BY C1, C2) DATA INITIALLY\n");
  printf("    DEFERRED REFRESH DEFERRED\n"); 
  
  /* directly execute the statement */
  cliRC = SQLExecDirect(hstmt, stmt2, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* create staging table */
  printf("\nCreating the staging table G\n");
  printf("\n  Directly execute the statement\n");
  printf("    CREATE TABLE G FOR D_AST PROPAGATE IMMEDIATE\n");
  
  /* directly execute the statement */
  cliRC = SQLExecDirect(hstmt, stmt3, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  return rc;
} /* CreateStagingTable */

/* show how to propagate the changes from base table to
   summary tables through the staging table */ 
int PropagateStagingToAst(SQLHANDLE hdbc)
{
  SQLRETURN cliRC = SQL_SUCCESS;
  int rc = 0;
  SQLHANDLE hstmt; /* statement handle */

  SQLCHAR *stmt1 = (SQLCHAR *)"SET INTEGRITY FOR G IMMEDIATE CHECKED ";

  SQLCHAR *stmt2 = (SQLCHAR *)"REFRESH TABLE D_AST NOT INCREMENTAL ";

  SQLCHAR *stmt3 = (SQLCHAR *)"INSERT INTO T VALUES(1,1,1,1), (2,2,2,2), "
                              "                    (1,1,1,1), (3,3,3,3)";
  
  SQLCHAR *stmt4 = (SQLCHAR *)"REFRESH TABLE D_AST INCREMENTAL" ;

  printf("\nUSE THE CLI FUNCTIONS\n");
  printf("  SQLSetConnectAttr\n");
  printf("  SQLAllocHandle\n");
  printf("  SQLExecDirect\n");
  printf("  SQLFreeHandle\n");
  printf("TO PROPAGATE THE CHANGES FROM BASE TABLE TO SUMMARY TABLE:\n");

  /* set AUTOCOMMIT on */
  cliRC = SQLSetConnectAttr(hdbc,
                            SQL_ATTR_AUTOCOMMIT,
                            (SQLPOINTER)SQL_AUTOCOMMIT_ON,
                            SQL_NTS);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  printf("\nBring staging table out of pending state\n");
  printf("\n  Directly execute the statement\n");
  printf("    SET INTEGRITY FOR G IMMEDIATE CHECKED\n");   
    
  /* directly execute the statement */
  cliRC = SQLExecDirect(hstmt, stmt1, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);
  
  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  printf("\nRefresh summary table, get it out of pending state.\n"); 
  printf("\n  Directly execute the statement\n");
  printf("    REFRESH TABLE D_AST NOT INCREMENTAL\n"); 
  
  /* directly execute the statement */
  cliRC = SQLExecDirect(hstmt, stmt2, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  printf("\nInsert data into base table T\n");
  printf("\n  Directly execute the statement\n");
  printf("    INSERT INTO T VALUES(1,1,1,1), (2,2,2,2), \n");
  printf("                        (1,1,1,1), (3,3,3,3)\n\n");
  
  /* directly execute the statement */
  cliRC = SQLExecDirect(hstmt, stmt3, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);    

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  printf("Display the contents of staging table G.\n"); 
  printf("The Staging table contains incremental changes to base table.\n"); 
  DisplayTable(hdbc, "G");
  
  printf("\nRefresh the summary table\n");
  printf("\n  Directly execute the statement\n");
  printf("    REFRESH TABLE D_AST INCREMENTAL\n");  
  
  /* directly execute the statement */
  cliRC = SQLExecDirect(hstmt, stmt4, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);    

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  printf("\nDisplay the contents of staging table G\n"); 
  printf("   NOTE: The staging table is pruned after AST is\n"); 
  printf("         refreshed. The contents are propagated to AST\n");
  printf("         from the staging table\n");
  DisplayTable(hdbc, "G");

  printf("Display the contents of AST\n");
  printf("Summary table has the changes propagated from staging table\n");
  DisplayTable(hdbc, "D_AST");
 
  return rc;
} /* PropageStagingToAst */

/* shows how to restore the data in a summary table */ 
int RestoreSummaryTable(SQLHANDLE hdbc)
{
  SQLRETURN cliRC = SQL_SUCCESS;
  int rc = 0;
  SQLHANDLE hstmt; /* statement handle */

  SQLCHAR *stmt1 = (SQLCHAR *) 
    "SET INTEGRITY FOR G OFF"; 

  SQLCHAR *stmt2 = (SQLCHAR *) 
    "SET INTEGRITY FOR D_AST OFF CASCADE IMMEDIATE ";

  SQLCHAR *stmt3 = (SQLCHAR *) 
    "SET INTEGRITY FOR G IMMEDIATE CHECKED PRUNE"; 

  SQLCHAR *stmt4 = (SQLCHAR *) 
    "SET INTEGRITY FOR G STAGING IMMEDIATE UNCHECKED";

  SQLCHAR *stmt5 = (SQLCHAR *) 
    "SET INTEGRITY FOR D_AST MATERIALIZED QUERY IMMEDIATE UNCHECKED"; 
  
  printf("\nUSE THE CLI FUNCTIONS\n");
  printf("  SQLSetConnectAttr\n");
  printf("  SQLAllocHandle\n");
  printf("  SQLExecDirect\n");
  printf("  SQLFreeHandle\n");
  printf("TO RESTORE DATA IN A SUMMARY TABLE:\n");

  /* set AUTOCOMMIT on */
  cliRC = SQLSetConnectAttr(hdbc,
                            SQL_ATTR_AUTOCOMMIT,
                            (SQLPOINTER)SQL_AUTOCOMMIT_ON,
                            SQL_NTS);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  printf("\nBlock all modifications to the summary table\n");
  printf("by setting the integrity to off\n");
  printf("  (G is placed in pending and G.CC=N)\n");
  printf("\n  Directly execute the statement\n");
  printf("    SET INTEGRITY FOR G OFF\n");
  
  /* directly execute the statement */
  cliRC = SQLExecDirect(hstmt, stmt1, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);    

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  printf("\nExport the query definition in summary table and load\n");
  printf("directly back to the summary table.\n");
  printf("  (D_AST and G both in pending)\n");
  printf("\n  Directly execute the statement\n");
  printf("    SET INTEGRITY FOR D_AST OFF CASCADE IMMEDIATE\n");

  /* directly execute the statement */
  cliRC = SQLExecDirect(hstmt, stmt2, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);    

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  printf("\nPrune staging table and place it in normal state\n");
  printf("  (G.CC=F)\n");
  printf("\n  Directly execute the statement\n");
  printf("    SET INTEGRITY FOR G IMMEDIATE CHECKED PRUNE\n"); 
  
  /* directly execute the statement */
  cliRC = SQLExecDirect(hstmt, stmt3, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);    
 
  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  printf("\nChanging staging table state to U\n");
  printf("  (G.CC to U)\n");
  printf("\n  Directly execute the statement\n");
  printf("    SET INTEGRITY FOR G STAGING IMMEDIATE UNCHECKED\n");

  /* directly execute the statement */
  cliRC = SQLExecDirect(hstmt, stmt4, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);   

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  printf("\nPlace D_AST in normal and D_AST.CC to U\n");
  printf("\n  Directly execute the statement\n");
  printf("    SET INTEGRITY FOR D_AST MATERIALIZED QUERY\n"); 
  printf("    IMMEDIATE UNCHECKED\n");

  /* directly execute the statement */
  cliRC = SQLExecDirect(hstmt, stmt5, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);   

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  return rc;
} /* RestoreSummaryTable */ 

/* displays the contents of the table being passed as the argument */
int DisplayTable(SQLHANDLE hdbc, char *table)
{
  SQLRETURN cliRC = SQL_SUCCESS;
  int rc = 0;
  SQLHANDLE hstmt; /* statement handle */
  
  char tbl[10];

  struct
  {
    SQLINTEGER ind;
    SQLSMALLINT val;
  }
  c1; /* variable to get data from the c1 column */

  struct
  {
    SQLINTEGER ind;
    SQLSMALLINT val;
  }
  c2; /* variable to get data from the c2 column */

  struct
  {
    SQLINTEGER ind;
    SQLSMALLINT val;
  }
  count; /* variable to get data from the count column */
  
  /* SQL SELECT statement to be executed */
  SQLCHAR *stmt1 = (SQLCHAR *)"SELECT c1, c2, count FROM G";

  SQLCHAR *stmt2 = (SQLCHAR *) "SELECT c1, c2, count FROM D_AST" ;

  printf("\nUSE THE CLI FUNCTIONS\n");
  printf("  SQLSetConnectAttr\n");
  printf("  SQLAllocHandle\n");
  printf("  SQLExecDirect\n");
  printf("  SQLFetch\n");
  printf("  SQLGetData\n");
  printf("  SQLFreeHandle\n");
  printf("TO DISPLAY THE CONTENTS OF THE TABLE:\n\n");

  /* set AUTOCOMMIT on */
  cliRC = SQLSetConnectAttr(hdbc,
                            SQL_ATTR_AUTOCOMMIT,
                            (SQLPOINTER)SQL_AUTOCOMMIT_ON,
                            SQL_NTS);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);
  
  strcpy(tbl, table);

  if(!strcmp(tbl, "G"))
  {
    printf("  Directly execute the statement\n");
    printf("    SELECT c1, c2, count FROM G\n\n");
    
    cliRC = SQLExecDirect(hstmt, stmt1, SQL_NTS);
    STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);     

  }
  else if(!strcmp(tbl, "D_AST"))
  {
    printf("  Directly execute the statement\n");
    printf("    SELECT c1, c2, count FROM D_AST\n\n");

    cliRC = SQLExecDirect(hstmt, stmt2, SQL_NTS);
    STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);     
  }
  
  printf("  C1    C2    COUNT \n");
  printf("  ------------------\n");
  
  /* fetch each row, and display */
  cliRC = SQLFetch(hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  if (cliRC == SQL_NO_DATA_FOUND)

  {
    printf("\n  Data not found.\n");
  }

  while (cliRC != SQL_NO_DATA_FOUND)
  {
    /* use SQLGetData to get the results */
    /* get data from column 1 */
    cliRC = SQLGetData(hstmt,
                       1,
                       SQL_C_SHORT,
                       &c1.val,
                       0,
                       &c1.ind);
    STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

    /* use SQLGetData to get the results */
    /* get data from column 2 */
    cliRC = SQLGetData(hstmt,
                       2,
                       SQL_C_SHORT,
                       &c2.val,
                       0,
                       &c2.ind);
    STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

    /* use SQLGetData to get the results */
    /* get data from column 3 */
    cliRC = SQLGetData(hstmt,
                       3,
                       SQL_C_SHORT,
                       &count.val,
                       0,
                       &count.ind);
    STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

    printf("  %d     %d       %d  \n", c1.val, c2.val, count.val);

     /* fetch next row */
    cliRC = SQLFetch(hstmt);
    STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);
  } 
  
  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  printf("\n");

  return rc;
} /* DisplayTable */

/* Drops the staging table, summary table and base table */
int DropTables(SQLHANDLE hdbc)
{
  SQLRETURN cliRC = SQL_SUCCESS;
  int rc = 0;
  SQLHANDLE hstmt; /* statement handle */

  printf("Dropping a base table implicitly drops summary table defined\n");
  printf("on it which in turn cascades to dropping its staging table.\n");
  
  printf("\nUSE THE CLI FUNCTIONS\n");
  printf("  SQLSetConnectAttr\n");
  printf("  SQLAllocHandle\n");
  printf("  SQLExecDirect\n");
  printf("  SQLFreeHandle\n");
  printf("TO DROP A TABLE:\n\n");

  /* set AUTOCOMMIT on */
  cliRC = SQLSetConnectAttr(hdbc,
                            SQL_ATTR_AUTOCOMMIT,
                            (SQLPOINTER)SQL_AUTOCOMMIT_ON,
                            SQL_NTS);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  printf("  Directly execute the statement\n");
  printf("    DROP TABLE T\n");

  /* create base table  */
  cliRC = SQLExecDirect(
            hstmt,
            (UCHAR *)
            "DROP TABLE T",
            SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  return rc;
} /* DropTables */

File: java/jdbc/TbAST.java
//**************************************************************************
// (c) Copyright IBM Corp. 2007 All rights reserved.
// 
// The following sample of source code ("Sample") is owned by International 
// Business Machines Corporation or one of its subsidiaries ("IBM") and is 
// copyrighted and licensed, not sold. You may use, copy, modify, and 
// distribute the Sample in any form without payment to IBM, for the purpose of 
// assisting you in the development of your applications.
// 
// The Sample code is provided to you on an "AS IS" basis, without warranty of 
// any kind. IBM HEREBY EXPRESSLY DISCLAIMS ALL WARRANTIES, EITHER EXPRESS OR 
// IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. Some jurisdictions do 
// not allow for the exclusion or limitation of implied warranties, so the above 
// limitations or exclusions may not apply to you. IBM shall not be liable for 
// any damages you suffer as a result of using, copying, modifying or 
// distributing the Sample, even if IBM has been advised of the possibility of 
// such damages.
// *************************************************************************
//
// SOURCE FILE NAME: TbAST.java
//
// SAMPLE: How to use staging table for updating deferred AST 
//          
//         This sample:
//         1. Creates a refresh-deferred summary table 
//         2. Creates a staging table for this summary table 
//         3. Applies contents of staging table to AST
//         4. Restores the data in a summary table 
//
// SQL STATEMENTS USED:
//         CREATE SUMMARY TABLE
//         DROP
//         INSERT
//         REFRESH
//         SET INTEGRITY
//
// JAVA 2 CLASSES USED:
//         Statement
//         ResultSet
//
// Classes used from Util.java are:
//         Db
//         JdbcException
//
// Compile: the utility file and the source file with:
//		javac Util.java
//		javac <filename>.java
//
// Run: java <filename> [<db_name>] <username> <pwd>
//	or
//     java <filename> [<db_name>] <server_name> <port_num> <username> <pwd>
//
// OUTPUT FILE: TbAST.out (available in the online documentation)
// Output will vary depending on the JDBC driver connectivity used.
//**************************************************************************
//
// For more information on the sample programs, see the README file.
//
// For information on developing JDBC applications, see the Application
// Development Guide.
//
// For information on using SQL statements, see the SQL Reference.
//
// For the latest information on programming, compiling, and running DB2
// applications, visit the DB2 application development website at
//     http://www.software.ibm.com/data/db2/udb/ad
//**************************************************************************

import java.lang.*;
import java.sql.*;

class TbAST
{
  public static void main(String argv[])
  {
    try
    {
      Db db = new Db(argv);

      System.out.println();
      System.out.println(
        "THIS SAMPLE SHOWS THE USAGE OF STAGING TABLE TO UPDATE \n" + 
        "REFRESH DEFERRED AST AND RESTORE DATA IN A SUMMARY TABLE \n" +
        "\n-----------------------------------------------------------\n");

      // connect to database
      db.connect();

      // create a base table, summary table, staging table 
      createStagingTable(db.con);

      // to show the propagation of changes of base table to
      // summary tables through the staging table 
      System.out.println(
        "\n-----------------------------------------------------------\n" +
        "To show the propagation of changes from base table to \n" +
        "summary tables through the staging table: \n" );
      propagateStagingToAst(db.con);
    
      // to show restoring of data in a summary table 
      System.out.println(
        "\n------------------------------------------------------------ \n" +
        "To show restoring of data in a summary table");
      restoreSummaryTable(db.con);
  
      // drop the created tables
      System.out.println(
        "\n------------------------------------------------------------ \n" +
        "Drop the created tables");
      dropTables(db.con);
 
      // disconnect from 'sample' database
      db.disconnect();
    }
    catch (Exception e)
    {
      JdbcException jdbcExc = new JdbcException(e) ;
      jdbcExc.handle();
    }
  } // main

  // Creates base table, summary table and staging table 
  static void createStagingTable(Connection con)
  {
    try
    {
      Statement stmt = con.createStatement();
            
      //create base table
      System.out.println(
        "\nUSE THE SQL STATEMENT: \n" +
        "  CREATE TABLE \n" +
        "To create base table, summary table, staging table\n" +
        "\nCreating the base table t\n" +
        "  CREATE TABLE t \n" +
        "    (c1 SMALLINT NOT NULL,\n" +
        "     c2 SMALLINT NOT NULL, \n" +
        "     c3 SMALLINT, \n" +
        "     c4 SMALLINT)\n");       
      stmt.executeUpdate(
        "  CREATE TABLE t " +
        "    (c1 SMALLINT NOT NULL, " +
        "     c2 SMALLINT NOT NULL, " +
        "     c3 SMALLINT, " +
        "     c4 SMALLINT)");      
      System.out.println("  COMMIT");
      con.commit();
      
      // create summary table
      System.out.println(
        "\nCreating summary table d_ast \n" +
        "  CREATE SUMMARY TABLE d_ast AS\n" + 
        "    (SELECT c1, c2, COUNT(*) AS count\n" + 
        "      FROM t\n" +
        "      GROUP BY c1, c2)\n" +
        "    DATA INITIALLY DEFERRED\n" + 
        "    REFRESH DEFERRED \n");
      stmt.executeUpdate(
        "CREATE SUMMARY TABLE d_ast AS " +
        "  (SELECT c1, c2, COUNT(*) AS count " +
        "    FROM t " +
        "    GROUP BY c1, c2) " +
        "  DATA INITIALLY DEFERRED " +
        "  REFRESH DEFERRED ");
      System.out.println("  COMMIT");
      con.commit();
     
      // create staging table
      System.out.println(
        "\nCreating the staging table g \n" + 
        "  CREATE TABLE g FOR d_ast PROPAGATE IMMEDIATE");
      stmt.executeUpdate("CREATE TABLE g FOR d_ast PROPAGATE IMMEDIATE");
      System.out.println("\n  COMMIT");
      con.commit();
      
      stmt.close();
    }
    catch (Exception e)
    {
      JdbcException jdbcExc = new JdbcException(e) ;
      jdbcExc.handle();
    }
  } // createStagingTable
  
  // Show how to propagate the changes from base table to
  // summary tables through the staging table
  static void propagateStagingToAst(Connection con)
  {
    try
    {
      Statement stmt = con.createStatement();
     
      System.out.println(
        "Bring staging table out of pending state \n" + 
        "  SET INTEGRITY FOR g IMMEDIATE CHECKED");
      stmt.executeUpdate("SET INTEGRITY FOR g IMMEDIATE CHECKED");      
      con.commit();
      
      System.out.println(
        "\nRefresh summary table, get it out of pending state. \n" + 
        "  REFRESH TABLE d_ast NOT INCREMENTAL");
      stmt.executeUpdate("REFRESH TABLE d_ast NOT INCREMENTAL");     
      con.commit();
     
      System.out.println(
        "\nInsert data into base table T\n" +
        "  INSERT INTO t VALUES(1,1,1,1), \n" +
        "                      (2,2,2,2), \n" +
        "                      (1,1,1,1), \n" +
        "                      (3,3,3,3)");
      stmt.executeUpdate(
        "INSERT INTO t VALUES(1,1,1,1)," +
        "                    (2,2,2,2)," +
        "                    (1,1,1,1)," +
        "                    (3,3,3,3)");      
      con.commit();
      
      System.out.println(
        "\nDisplay the contents of staging table g.\n" + 
        "The Staging table contains incremental changes to base table.\n"); 
      displayTable(con, "g");
 
      System.out.println(
        "\n\nRefresh the summary table \n" +
        "  REFRESH TABLE d_ast INCREMENTAL");
      stmt.executeUpdate("REFRESH TABLE d_ast INCREMENTAL");
      con.commit();
      
      System.out.println(
        "\nDisplay the contents of staging table g \n" +
        "   NOTE: The staging table is pruned after AST is \n" +
        "         refreshed. The contents are propagated to AST \n" +
        "         from the staging table\n");
      displayTable(con, "g");

      System.out.println(
        "\nDisplay the contents of AST\n"  +
        "Summary table has the changes propagated from staging table\n");
      displayTable(con, "d_ast");
      
      stmt.close();
    }
    catch (Exception e)
    {
      JdbcException jdbcExc = new JdbcException(e) ;
      jdbcExc.handle();
    }
  } // propagateStagingToAst
  
  // Shows how to restore the data in a summary table 
  static void restoreSummaryTable(Connection con)
  {
    try
    {
      Statement stmt = con.createStatement();
            
      System.out.println(
        "\nBlock all modifications to the summary table \n" +
        "by setting the integrity to off \n" +
        "  (g is placed in pending and g.CC=N) \n" +
        "  SET INTEGRITY FOR g OFF");
      stmt.executeUpdate("SET INTEGRITY FOR g OFF");
      con.commit();
      
      System.out.println(
        "\nExport the query definition in summary table and load \n" +
        "directly back to the summary table.\n" +
        "  (d_ast and g both in pending \n" + 
        "  SET INTEGRITY FOR d_ast OFF CASCADE IMMEDIATE\n");
      stmt.executeUpdate("SET INTEGRITY FOR d_ast OFF CASCADE IMMEDIATE");
      con.commit();
                                           
      System.out.println(
        "Prune staging table and place it in normal state\n" +
        "  (g.CC=F)\n" + 
        "  SET INTEGRITY FOR g IMMEDIATE CHECKED PRUNE\n");
      stmt.executeUpdate("SET INTEGRITY FOR g IMMEDIATE CHECKED PRUNE");
      con.commit();
               
      System.out.println(
        "Changing staging table state to U \n"  +
        "  (g.CC to U)\n" + 
        "  SET INTEGRITY FOR g STAGING IMMEDIATE UNCHECKED");
      stmt.executeUpdate("SET INTEGRITY FOR g STAGING IMMEDIATE UNCHECKED");
      con.commit();
      
      System.out.println(
        "\nPlace d_ast in normal and d_ast.CC to U \n" + 
        "  SET INTEGRITY FOR d_ast MATERIALIZED QUERY IMMEDIATE UNCHECKED");
      stmt.executeUpdate(
        "SET INTEGRITY FOR d_ast MATERIALIZED QUERY IMMEDIATE UNCHECKED");
      con.commit();
      
      stmt.close();
    }
    catch (Exception e)
    {
      JdbcException jdbcExc = new JdbcException(e) ;
      jdbcExc.handle();
    }
  } // restoreSummaryTable
  
  // Displays the contents of the table being passed as the argument
  static void displayTable(Connection con, String tableName)
  {
    try
    {
      Statement stmt = con.createStatement();
      String sqlString = null;
      ResultSet rs;
      int c1 = 0;
      int c2 = 0;
      int count = 0;
      
      if (tableName.equals("g"))
      {
        sqlString = "  SELECT c1, c2, count FROM g" ;
      }
      else if (tableName.equals("d_ast"))
      {
        sqlString = "  SELECT c1, c2, count FROM d_ast" ;
      }
      
      rs = stmt.executeQuery(sqlString);
      System.out.println(
        sqlString +
        "\n\n  C1    C2    COUNT " +
        "\n  ------------------");
      
      while (rs.next())
      {
        c1 = rs.getInt("c1");
        c2 = rs.getInt("c2");
        count = rs.getInt("count");
         
        System.out.println("   " + c1 + "    " + c2 + "    " + count );
      }
      
      rs.close();
      stmt.close();
    }  
    catch (Exception e)
    {
      JdbcException jdbcExc = new JdbcException(e) ;
      jdbcExc.handle();
    }
  } // displayTable
  
  // Drops the staging table, summary table and base table
  static void dropTables(Connection con)
  { 
    try
    {
      Statement stmt = con.createStatement();
     
      System.out.println(
        "\nDropping a base table implicitly drops summary table defined \n" +
        "on it which in turn cascades to dropping its staging table. \n" +
        "\nUSE THE SQL STATEMENT:\n" +
        "  DROP TABLE \n" + 
        "To drop a table \n\n" + 
        "  DROP TABLE t \n");
      stmt.executeUpdate("DROP TABLE t");
      System.out.println("  COMMIT");
      con.commit();
      
      stmt.close();
    }
    catch (Exception e)
    {
      JdbcException jdbcExc = new JdbcException(e) ;
      jdbcExc.handle();
    }
  } // dropTables
} // TbAST
